src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java

```diff
@@ -831,7 +977,7 @@ public SCMRevision run(GitClient client, String remoteName) throws IOException,
                                       }
                                       listener.getLogger()
                                               .printf("Selected match: %s revision %s%n", name, hash);
-                                      return new SCMRevisionImpl(new SCMHead(name), hash);
+                                      return new GitBranchSCMRevision(new GitBranchSCMHead(name), hash);
```

Comment:
We are missing the case where the revision does not exist on any branch but does exist in the history of an other ref. Now if that is because `client.getBranchesContaining(hash, true)` doesn't allow us to query other refs, then we need to file a JIRA, add a comment and move on... otherwise we probably need to revisit the `candidatePrefix` search to include other refs

Reply:
Well, the way I think the refspec for the trait is constructed the commits will be fetched as branches on the remote, I don't like it, but it should work. I should probably add a test for that though to verify.